homeworks/week4/Huy_Anh.java

```diff
@@ -0,0 +1,171 @@
+                new Interval(1,3),
+                new Interval(2,5),
+                new Interval(1,4),
+                new Interval(6,8),
+        }; //Hybrid [1,5] U [6,8]
+
+        Interval[] intervals4 = {
+                new Interval(0,1),
+                new Interval(3,9),
+                new Interval(10, 13)
+        }; //Separated with different length [0,1] U [3,9] U [10,13]
+
+        Interval[] originalTest = {
+                new Interval(2,5),
+                new Interval(1,3)
+        };
+        test(intervals);
+        System.out.println("========");
+        test(originalTest);
+        System.out.println("========");
+
+        // Test for generateRandomValue function only.
+
+        //test(intervals2);
+        //System.out.println("========");
+        //test(intervals3);
+        //System.out.println("========");
+        //test(intervals4);
+
+    }
+
+    public static void test(Interval[] intervals){
+        double low = Integer.MAX_VALUE;
+        double high = Integer.MIN_VALUE;
+        for(Interval interval : intervals){
+            if(interval.low < low)
+                low = interval.low;
+            if(interval.high > high)
+                high = interval.high;
+        }
+        int length = (int)(high-low);
+        int[] count = new int[length];
+        int iterations = 100000;
+        for (int i = 0; i < iterations; i++) {
+            // Implemented from Mr Tuan Anh's explanation
+
+            //double f = generateRandomValue(intervals);
+
+            // Implemented from Mr Hien's explanation
+
+            double f = oldGenerateRandomValue(intervals);
+
+            for (int j = 0; j < length; j++) {
+                if (f >= (j + low) && f < (j + low + 1)) {
+                    count[j]++;
+                }
+            }
+        }
+
+        for (int j = 0; j < length; j++) {
+            double prob = 100*(double)count[j]/iterations;
+            if (prob > 0)
+                System.out.println("Interval [" + (int)(j + low) + "," + (int)(j + low + 1) + ") with probability " + prob + " %");
+        }
+    }
+    public static double oldGenerateRandomValue(Interval[] intervals) {
+        if (intervals == null || intervals.length == 0) return -1;
+        Random rnd = new Random();
+        double rndNumber = rnd.nextDouble(); // Uniformly generate a number
+        double firstRange = 0;
+        double span = 0;
+        for (Interval interval : intervals) // O(n) Time
+            span += interval.length();
+        for (Interval interval : intervals){ // O(n) Time
+            if(rndNumber >= firstRange/span && rndNumber < (firstRange+interval.length())/span){
+                return rndNumber*span - (firstRange - interval.low);
+            }
+            firstRange += interval.length();
+        }
+        return -1;
+    }
+    // Time Complexity: O(2kn) with k is sample times. It can be reduced to O(kn) + O(n) if taking the loop to calculate span outside the function.
```

Comment:
Since generating random values are independent events, you don't have to take `k` into account when calculating time/space complexity. But it's nice that you think of it.